ports/esp32/machine_pwm.c

```diff
@@ -360,47 +525,76 @@ STATIC mp_obj_t mp_machine_pwm_make_new(const mp_obj_type_t *type,
+                timers[chans[channel_idx].timer_idx].freq_hz = -1;
+            }
         }
 
-        // Mark it unused, and tell the hardware to stop routing
-        check_esp_err(ledc_stop(self->mode, chan, 0));
-        // Disable ledc signal for the pin
-        // gpio_matrix_out(self->pin, SIG_GPIO_OUT_IDX, false, false);
-        if (self->mode == LEDC_LOW_SPEED_MODE) {
-            gpio_matrix_out(self->pin, LEDC_LS_SIG_OUT0_IDX + self->channel, false, true);
-        } else {
-            #if LEDC_SPEED_MODE_MAX > 1
-            #if CONFIG_IDF_TARGET_ESP32
-            gpio_matrix_out(self->pin, LEDC_HS_SIG_OUT0_IDX + self->channel, false, true);
-            #else
-            #error Add supported CONFIG_IDF_TARGET_ESP32_xxx
-            #endif
-            #endif
+        int pin = chans[channel_idx].pin;
+        if (pin != -1) {
+            int mode = CHANNEL_IDX_TO_MODE(channel_idx);
+            int channel = CHANNEL_IDX_TO_CHANNEL(channel_idx);
+            // Mark it unused, and tell the hardware to stop routing
+            check_esp_err(ledc_stop(mode, channel, 0));
+            // Disable ledc signal for the pin
+            // gpio_matrix_out(pin, SIG_GPIO_OUT_IDX, false, false);
+            if (mode == LEDC_LOW_SPEED_MODE) {
+                gpio_matrix_out(pin, LEDC_LS_SIG_OUT0_IDX + channel, false, true);
+            } else {
+                #if LEDC_SPEED_MODE_MAX > 1
+                #if CONFIG_IDF_TARGET_ESP32
+                gpio_matrix_out(pin, LEDC_HS_SIG_OUT0_IDX + channel, false, true);
+                #else
+                #error Add supported CONFIG_IDF_TARGET_ESP32_xxx
+                #endif
+                #endif
+            }
+        }
+        chans[channel_idx].pin = -1;
+        chans[channel_idx].timer_idx = -1;
+    }
+}
+
+// This called from pwm.deinit() method
+STATIC void mp_machine_pwm_deinit(machine_pwm_obj_t *self) {
+    int channel_idx = CHANNEL_IDX(self->mode, self->channel);
+    pwm_deinit(channel_idx);
+    self->active = false;
+    self->mode = -1;
+    self->channel = -1;
+    self->timer = -1;
+    self->duty_x = 0;
+}
+
+// This called from Ctrl-D soft reboot
+void machine_pwms_deinit(void) {
+    if (pwm_inited) {
+        for (int channel_idx = 0; channel_idx < PWM_CHANNEL_MAX; ++channel_idx) {
+            pwm_deinit(channel_idx);
         }
-        chans[chan].pin = -1;
-        chans[chan].timer_idx = -1;
-        self->active = false;
-        self->mode = -1;
-        self->channel = -1;
-        self->timer = -1;
+        pwm_inited = false;
     }
 }
 
+// Set's and get's methods of PWM class
+
 STATIC mp_obj_t mp_machine_pwm_freq_get(machine_pwm_obj_t *self) {
     return MP_OBJ_NEW_SMALL_INT(ledc_get_freq(self->mode, self->timer));
 }
 
 STATIC void mp_machine_pwm_freq_set(machine_pwm_obj_t *self, mp_int_t freq) {
+    if ((freq <= 0) || (freq > 40000000)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("freqency must be from 1Hz to 40MHz"));
```

Comment:
"frequency must be between 1Hz and 40MHz"

Reply:
I asked here 
https://github.com/micropython/micropython/pull/6639#issuecomment-944587837
and got two answers with 
**"between" is most of the time excluding the border**

I prefer more clear
**from 1Hz to 40MHz** which includes the border.